cmd/crossplane/main.go

```diff
@@ -38,106 +40,166 @@ import (
-		extManageTemplatesController = extManageCmd.Flag("templating-controller-image", "The image of the Template Stacks controller (implies --templates)").Default("").String()
-
-		extManageHostControllerNamespace = extManageCmd.Flag("host-controller-namespace", "The namespace on Host Cluster where install and controller jobs/deployments will be created. Setting this will activate host aware mode of Stack Manager").String()
-		extManageTenantKubeconfig        = extManageCmd.Flag("tenant-kubeconfig", "The absolute path of the kubeconfig file to Tenant Kubernetes instance (required for host aware mode, ignored otherwise).").ExistingFile()
-
-		// Unpack the given stack package content. This command is expected to
-		// parse the content and generate manifests for stack related artifacts
-		// to stdout so that the SM can read the output and use the Kubernetes
-		// API to create the artifacts.
-		//
-		// Users are not expected to run this command themselves, the stack
-		// manager itself should execute this command.
-		//
-		// Unpack does not interact with the Kubernetes API.
-		extUnpackCmd                 = extCmd.Command("unpack", "Unpack a Stack").Alias("unstack")
-		extUnpackDir                 = extUnpackCmd.Flag("content-dir", "The absolute path of the directory that contains the stack contents").Required().String()
-		extUnpackOutfile             = extUnpackCmd.Flag("outfile", "The file where the YAML Stack record and CRD artifacts will be written").String()
-		extUnpackPermissionScope     = extUnpackCmd.Flag("permission-scope", "The permission-scope that the stack must request (Namespaced, Cluster)").Default("Namespaced").String()
-		extUnpackTemplatesController = extUnpackCmd.Flag("templating-controller-image", "The image of the Template Stacks controller").Default("").String()
-	)
-	cmd := kingpin.MustParse(app.Parse(os.Args[1:]))
-
-	zl := zap.New(zap.UseDevMode(*debug))
+	baseCmd := filepath.Base(os.Args[0])
+	args := os.Args[1:]
+
+	app := kingpin.New(baseCmd, "An open source multicloud control plane.").DefaultEnvars()
+	debug := app.Flag("debug", "Run with debug logging.").Short('d').Bool()
+	syncPeriod := app.Flag("sync", "Controller manager sync period duration such as 300ms, 1.5h or 2h45m").Short('s').Default("1h").Duration()
+
+	a := appReq{
+		app:        app,
+		zl:         zap.New(zap.UseDevMode(*debug)),
+		syncPeriod: syncPeriod,
+	}
+
+	commands := []fullCommand{}
+	commands = append(commands, a.crossplaneCmd(baseCmd)...)
+	commands = append(commands, a.stackCmd()...)
+
+	cmdGiven := kingpin.MustParse(app.Parse(args))
+
 	if *debug {
 		// The controller-runtime runs with a no-op logger by default. It is
 		// *very* verbose even at info level, so we only provide it a real
 		// logger when we're running in debug mode.
-		ctrl.SetLogger(zl)
+		ctrl.SetLogger(a.zl)
+	}
+
+	knownCmd := false
+	for _, cmd := range commands {
```

Comment:
How did we decide to take this approach, rather than have a `case` statement? On one hand I like that it makes it easy to add new commands and handle them in the same way, but on the other hand, I'm not sure about having the `break`.

Reply:
Because the commands are a slice I can't offer a switch over an unknown quantity of command cases.

The `for` loop was envisioned while breaking up the cyclomatically complex main.  Now that it has been broken up I can see how this could be turned back into a switch statement.  That said, we have three commands to process now.  If this switch statement doesn't trigger the complexity warnings the next command (or the next one) could do it and we would be revisiting this again.